internals-js/src/specs/sourceSpec.ts

```diff
@@ -140,8 +144,271 @@ export class SourceSpecDefinition extends FeatureDefinition {
+          }
+        } else {
+          errors.push(new GraphQLError(
+            `${sourceAPIInSchema.name} must specify one of ${KNOWN_SOURCE_PROTOCOLS.join(', ')}`,
+          ));
+        }
+      });
+    }
+
+    if (sourceTypeInSchema) {
+      sourceTypeInSchema.applications().forEach(application => {
+        const { api, ...rest } = application.arguments();
+        if (!api || !apiNameToProtocol.has(api)) {
+          errors.push(new GraphQLError(`${sourceTypeInSchema.name} specifies unknown api ${api}`));
+        } else {
+          const expectedProtocol = apiNameToProtocol.get(api);
+          const protocolValue = expectedProtocol && rest[expectedProtocol];
+          if (expectedProtocol && !protocolValue) {
+            errors.push(new GraphQLError(
+              `${sourceTypeInSchema.name} must specify same ${
+                expectedProtocol
+              } argument as corresponding ${
+                sourceAPIInSchema!.name
+              } for api ${api}`,
+            ));
+          }
+
+          if (protocolValue && expectedProtocol === "http") {
+            const { GET, POST, headers, body } = protocolValue as HTTPSourceType;
+
+            if ([GET, POST].filter(Boolean).length !== 1) {
+              errors.push(new GraphQLError(
+                `${sourceTypeInSchema.name} must specify exactly one of http.GET or http.POST`,
+              ));
+            } else {
+              const urlPathTemplate = (GET || POST)!;
+              try {
+                // TODO Validate URL path template uses only available fields of
+                // the type.
+                parseURLPathTemplate(urlPathTemplate);
+              } catch (e) {
+                errors.push(new GraphQLError(
+                  `${sourceTypeInSchema.name} http.GET or http.POST must be valid URL path template`,
+                ));
+              }
+            }
+
+            validateHTTPHeaders(headers, errors, sourceTypeInSchema.name);
+
+            if (body) {
+              try {
+                parseJSONSelection(body);
+                // TODO Validate body selection matches the available fields.
+              } catch (e) {
+                errors.push(new GraphQLError(
+                  `${sourceTypeInSchema.name} http.body not valid JSONSelection: ${
+                    e.message
+                  }`,
+                ));
+              }
+            }
+          }
+        }
+
+        const ast = application.parent.sourceAST;
+        switch (ast?.kind) {
+          case "ObjectTypeDefinition":
+          case "InterfaceTypeDefinition":
+            if (!ast.directives?.some(directive => directive.name.value === "key")) {
+              errors.push(new GraphQLError(
+                `${sourceTypeInSchema.name} must be applied to an entity type that also has a @key directive`,
+              ));
+            }
+            // TODO Validate selection is valid JSONSelection for type.
```

Comment:
So here we are going to validate that the key values match the inputs to the URLPathTemplate? Isn't that `body` rather than `selection`?

Reply:
The `selection` argument for `@sourceType` needs to translate from REST JSON to GraphQL fields. I'm checking that here because this code theoretically has easy access to the `ObjectTypeDefinition` or `InterfaceTypeDefinition` to compare its fields to the output shape of the `selection` string.